source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc

```diff
@@ -0,0 +1,1109 @@
+  F(binaryAnnotationsField, binary_annotations)                                                    \
+  F(debugField, debug)
+
+#define ANNOTATION_FIELD_NAMES(F)                                                                  \
+  F(timestampField, timestamp)                                                                     \
+  F(valueField, value)                                                                             \
+  F(hostField, host)                                                                               \
+  F(keyField, key)                                                                                 \
+  F(annotationTypeField, annotation_type)
+
+#define ENDPOINT_FIELD_NAMES(F)                                                                    \
+  F(ipv4Field, ipv4)                                                                               \
+  F(portField, port)                                                                               \
+  F(serviceNameField, service_name)
+
+STRUCT_NAMES(MAKE_STATIC_STRING)
+REQUEST_HEADER_FIELD_NAMES(MAKE_STATIC_STRING)
+CLIENT_ID_FIELD_NAMES(MAKE_STATIC_STRING)
+DELEGATION_FIELD_NAMES(MAKE_STATIC_STRING)
+RESPONSE_HEADER_FIELD_NAMES(MAKE_STATIC_STRING)
+SPAN_FIELD_NAMES(MAKE_STATIC_STRING)
+ANNOTATION_FIELD_NAMES(MAKE_STATIC_STRING)
+ENDPOINT_FIELD_NAMES(MAKE_STATIC_STRING)
+
+const std::string& emptyString() { CONSTRUCT_ON_FIRST_USE(std::string, ""); }
+
+/**
+ * HeaderObjectProtocol implements BinaryProtocolImpl for the specific purpose of decoding the
+ * Twitter protocol RequestHeader and ResponseHeader thrift structs. These appear after any
+ * transport data (e.g. frame size) and before the start of a Thrift message. Decoding them
+ * via a Protocol implementation allows us to reuse the Decoder and its state machine.
+ */
+class HeaderObjectProtocol : public BinaryProtocolImpl {
+public:
+  bool readMessageBegin(Buffer::Instance&, MessageMetadata&) override { return true; }
+  bool readMessageEnd(Buffer::Instance&) override { return true; }
+};
+
+// Not const because the interfaces do not allow it, but these objects do not maintain internal
+// state and are therefore not modifiable.
+Transport& headerObjectTransport() {
+  static UnframedTransportImpl* transport = new UnframedTransportImpl();
+  return *transport;
+}
+
+Protocol& headerObjectProtocol() {
+  static HeaderObjectProtocol* protocol = new HeaderObjectProtocol();
+  return *protocol;
+}
+
+/**
+ * ClientId is a Twitter protocol client identifier.
+ *
+ * See https://github.com/twitter/finagle/blob/master/finagle-thrift/src/main/thrift/tracing.thrift
+ */
+class ClientId {
+public:
+  ClientId(const std::string& name) : name_(name) {}
+  ClientId(const ThriftStructValue& value) {
+    for (const auto& field : value.fields()) {
+      if (field->fieldId() == NameFieldId) {
+        name_ = field->getValue().getValueTyped<std::string>();
```

Comment:
should we log when we see more than one client id (e.g.: `name_` gets set more than once)? also, should we document why we keep the last occurence?

Reply:
There isn't any provision for sending multiple client ids. The loop here is over the fields of the ClientId struct (which has only ever had a single field, with default requiredness). The struct itself is optional as well.